### PR TITLE
Subaru: Global gen1 experimental longitudinal

### PR DESCRIPTION
**Pull request by jnewb1**
- Experimental long for gen1 subarus, gated behind an experimental flag for testing

### CODE REVIEW

**Comment by jnewb1** [comment]
I'd like a way to run the longitudinal tests without having to enable experimentalLongAvailable, will look more into that

**Comment by jnewb1** [comment]
c3d1ccb52f5f9d65|2023-08-16--10-27-16--3

about 20 minutes of running on our outback with a gen1 simulator (2 extra pandas that put everything on the right bus to make it look like gen1 to OP and gen2 to the car)

got a fault after an AEB triggered, so we will have work around that

**Comment by adeebshihadeh** [comment]
> got a fault after an FCW triggered, so we will have work around that

Doesn't matter, support code should be merged first.

File: selfdrive/car/subaru/values.py [hunk 1]
```diff
@@ -29,9 +29,33 @@ def __init__(self, CP):
...
+  RPM_LOOKUP_BP = [0, 1]
+  RPM_LOOKUP_V = [RPM_INACTIVE, RPM_MAX]
+
+  BRAKE_LOOKUP_BP = [-1, 0]
+  BRAKE_LOOKUP_V = [BRAKE_MAX, BRAKE_MIN]
+
 
 class SubaruFlags(IntFlag):
   SEND_INFOTAINMENT = 1
+  EXPERIMENTAL_LONG = 2 # EXTREMELY experimental
```
Comment (adeebshihadeh) [inline selfdrive/car/subaru/values.py:58]: why is this needed? should be able to use CP.openpilotLongitudinalControl

File: selfdrive/car/subaru/carcontroller.py [hunk 2]
```diff
@@ -20,6 +21,7 @@ def update(self, CC, CS, now_nanos):
     actuators = CC.actuators
     hud_control = CC.hudControl
     pcm_cancel_cmd = CC.cruiseControl.cancel
+    long_enabled = self.CP.openpilotLongitudinalControl
```
Comment (adeebshihadeh) [inline selfdrive/car/subaru/carcontroller.py:24]: I see it's already done above, but I'd avoid doing this. it just introduces an unnecessary indirection

File: selfdrive/car/subaru/interface.py [hunk 3]
```diff
@@ -107,6 +107,21 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
     else:
       raise ValueError(f"unknown car: {candidate}")
 
+    #ret.flags |= SubaruFlags.EXPERIMENTAL_LONG.value
+    experimental_long_allowed = (ret.flags & SubaruFlags.EXPERIMENTAL_LONG) != 0
+
+    # longitudinal
+    ret.experimentalLongitudinalAvailable = experimental_long_allowed and candidate not in (GLOBAL_GEN2 | PREGLOBAL_CARS | LKAS_ANGLE)
+    if experimental_long and ret.experimentalLongitudinalAvailable:
```
Comment (adeebshihadeh) [inline selfdrive/car/subaru/interface.py:115]:
```suggestion
    if ret.openpilotLongitudinalControl:
```

File: selfdrive/car/subaru/interface.py [hunk 4]
```diff
@@ -107,6 +107,21 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
     else:
       raise ValueError(f"unknown car: {candidate}")
 
+    #ret.flags |= SubaruFlags.EXPERIMENTAL_LONG.value
+    experimental_long_allowed = (ret.flags & SubaruFlags.EXPERIMENTAL_LONG) != 0
+
+    # longitudinal
+    ret.experimentalLongitudinalAvailable = experimental_long_allowed and candidate not in (GLOBAL_GEN2 | PREGLOBAL_CARS | LKAS_ANGLE)
```
Comment (adeebshihadeh) [inline selfdrive/car/subaru/interface.py:114]: make sure to comment this out before merging

File: selfdrive/car/subaru/subarucan.py [hunk 5]
```diff
@@ -138,12 +147,68 @@ def create_es_dashstatus(packer, dashstatus_msg):
...
+    "Brake_Lights",
+    "Cruise_Hold",
+    "Signal3",
+  ]}
+
+  if long_enabled:
+    values["Cruise_RPM"] = cruise_rpm
+
+  if long_active:
+    values["Cruise_Activated"] = 1
```
Comment (adeebshihadeh) [inline selfdrive/car/subaru/subarucan.py:208]: I'd indent this

File: selfdrive/car/subaru/values.py (hunk 1 above)
Comment (jnewb1) [reply selfdrive/car/subaru/values.py]: done

File: selfdrive/car/subaru/carcontroller.py (hunk 2 above)
Comment (jnewb1) [reply selfdrive/car/subaru/carcontroller.py]: done

File: selfdrive/car/subaru/interface.py (hunk 3 above)
Comment (jnewb1) [reply selfdrive/car/subaru/interface.py]: done

File: selfdrive/car/subaru/interface.py (hunk 4 above)
Comment (jnewb1) [reply selfdrive/car/subaru/interface.py]: done

File: selfdrive/car/subaru/subarucan.py (hunk 5 above)
Comment (jnewb1) [reply selfdrive/car/subaru/subarucan.py]: done
